### PR TITLE
Fix fonts are truncated problem

### PR DESCRIPTION
**Pull request by hykwok**
Calculate the additional height for the text area to solve font truncated problem (issue #46)

### CODE REVIEW

**Comment by jleclanche** [comment]
Can you please indent consistently?

**Comment by jleclanche** [comment]
Also you need to fix your commit author and name. And this is a very hacky way of fixing the issue, not a real fix.

**Comment by hykwok** [comment]
Just for the information. (Sorry for my poor English)

The height of the "drawing area" of a text is retrieved from QFontMetrics class.
According to the Qt document, the height is equal to "ascent + descent + 1".
So, normally, the calculated value for the height should cover the whole text.
However, during my testing, in different cases (different fonts in same size or same font in different sizes),the vertical position of the text was shifted.
That means, the area did not fully cover the "output" of the text and finally the text was truncated in the display.

I used some painting tools to study the output of the text for different fonts and found the baseline of the those fonts are different.
I also checked the results of the "ascent" and "descent" and found that, in some cases, the results were not correct. (For example, the value of ascent is equal to the descent.....)
Also, in the Qt document, 

1. int QFontMetrics::ascent() const
Returns the ascent of the font.
The ascent of a font is the distance from the baseline to the highest position characters extend to. In practice, some font designers break this rule, e.g. when they put more than one accent on top of a character, or to accommodate an unusual character in an exotic language, so it is possible (though rare) that this value will be too small.

2. int QFontMetrics::descent() const
Returns the descent of the font.
The descent is the distance from the base line to the lowest point characters extend to. In practice, some font designers break this rule, e.g. to accommodate an unusual character in an exotic language, so it is possible (though rare) that this value will be too small.

That means QFontMetrics class will return incorrect values for some fonts.

Therefore, I believe that we cannot simply use the returned value of the height for the calculation of the "drawing area".

Actually, if you check the "boundingRect" result of the QPainter::drawText (the actual bounding rectangle of the output, see note *1), the height of the "actual" output is larger than the height of the font (ascent + descent + 1).

I did not study the Qt source codes so I do not know how Qt renders the font during text drawing.
However, I found that the "boundingRect" result is a useful information for us to calculate the "drawing area".
The drawback is we need to call "QPainter::drawText" once during font switching in order to get the "height" of the "drawing area".

On the other hand, I think this bug is really serious because we cannot get the "complete" information from the display.
In my opinion, this widget is useless if this bug is not solved.

I cannot make sure the fix can solve the problem completely for all fonts but at least it works for my application.

Note:
*1:
void QPainter::drawText(int x, int y, int width, int height, int flags, const QString & text, QRect * boundingRect = 0)
Draws the given text within the rectangle with origin (x, y), width and height.
The boundingRect (if not null) is set to the actual bounding rectangle of the output.

**Comment by Mischanix** [comment]
I was thinking about this more and it does seem like this is the only way (definitely the simplest way)... it's not that wasteful resource wise, as long as Qt doesn't do any funny business.

I would just want the indentation to be consistent in the lines you change (I know it's not entirely consistent across the project at the moment, but it is at least locally consistent), and that the commit have a real author (i.e. with your actual email)

File: lib/TerminalDisplay.cpp [hunk 1]
```diff
@@ -242,9 +242,32 @@ void TerminalDisplay::fontChange(const QFont&)
...
+  // Although this operation will destory the orignal content, 
+  // the content will be drawn again after the test.
+  _drawTextTestFlag = true;
   update();
 }
 
+void TerminalDisplay::calDrawTextAdditionHeight(QPainter& painter)
+{    
+    QRect test_rect, feedback_rect;
+	test_rect.setRect(1, 1, _fontWidth * 4, _fontHeight);
```
Comment (Mischanix) [inline lib/TerminalDisplay.cpp]: Use spaces, not tabs.

File: lib/TerminalDisplay.cpp [hunk 2]
```diff
@@ -242,9 +242,32 @@ void TerminalDisplay::fontChange(const QFont&)
...
   update();
 }
 
+void TerminalDisplay::calDrawTextAdditionHeight(QPainter& painter)
+{    
+    QRect test_rect, feedback_rect;
+	test_rect.setRect(1, 1, _fontWidth * 4, _fontHeight);	
+	painter.drawText(test_rect, Qt::AlignBottom, LTR_OVERRIDE_CHAR + QString("Mq"), &feedback_rect);
+	
+	//qDebug() << "test_rect:" << test_rect << "feeback_rect:" << feedback_rect;
```
Comment (Mischanix) [inline lib/TerminalDisplay.cpp]: Remove debug code

File: lib/TerminalDisplay.cpp [hunk 3]
```diff
@@ -729,12 +756,15 @@ void TerminalDisplay::drawCharacters(QPainter& painter,
     // This was discussed in: http://lists.kde.org/?t=120552223600002&r=1&w=2
         if (_bidiEnabled)
             painter.drawText(rect,0,text);
-        else
+        else {
 #if QT_VERSION >= 0x040800
-            painter.drawText(rect, Qt::AlignBottom, LTR_OVERRIDE_CHAR + text);
+            QRect drawRect(rect.topLeft(), rect.size());
```
Comment (Mischanix) [inline lib/TerminalDisplay.cpp]: Haven't tested myself, but could you fix this issue for all versions by using drawRect on both sides of the #if here?

File: lib/TerminalDisplay.h [hunk 4]
```diff
@@ -700,6 +700,8 @@ private slots:
     void makeImage();
 
     void paintFilters(QPainter& painter);
+	
+	void calDrawTextAdditionHeight(QPainter& painter);
```
Comment (Mischanix) [inline lib/TerminalDisplay.h]: Name this `updateDrawPadding`.

File: lib/TerminalDisplay.h [hunk 5]
```diff
@@ -726,6 +728,8 @@ private slots:
     int  _fontWidth;     // width
     int  _fontAscent;     // ascend
     bool _boldIntense;   // Whether intense colors should be rendered with bold font
+	int  _drawTextAdditionHeight;	// additional height to prevent font trancation
```
Comment (Mischanix) [inline lib/TerminalDisplay.h]: Name this `drawPaddingHeight`

File: lib/TerminalDisplay.h [hunk 6]
```diff
@@ -726,6 +728,8 @@ private slots:
...
+	bool _drawTextTestFlag; 		// indicate it is a testing or not
```
Comment (Mischanix) [inline lib/TerminalDisplay.h]: Name this `needUpdateDrawPadding`

File: lib/TerminalDisplay.cpp [hunk 7]
```diff
@@ -1230,15 +1260,19 @@ void TerminalDisplay::focusInEvent(QFocusEvent*)
...
-
-  foreach (const QRect &rect, (pe->region() & contentsRect()).rects())
-  {
-    drawBackground(paint,rect,palette().background().color(),
-                    true /* use opacity setting */);
-    drawContents(paint, rect);
+  
+  if(_drawTextTestFlag) {
+      calDrawTextAdditionHeight(paint);      
+  } else {
```
Comment (Mischanix) [inline lib/TerminalDisplay.cpp]: Doesn't this else block prevent the paintEvent from painting what it should have if the draw padding needed an update?  If so, remove the else block.

File: lib/TerminalDisplay.cpp [hunk 8]
```diff
@@ -340,6 +363,10 @@ TerminalDisplay::TerminalDisplay(QWidget *parent)
 ,_cursorShape(BlockCursor)
 ,mMotionAfterPasting(NoMoveScreenWindow)
 {
+  // variables for draw text
```
Comment (Mischanix) [inline lib/TerminalDisplay.cpp]: Keep with the pattern set above and use the constructor member initializer syntax for these initializations.

File: lib/TerminalDisplay.cpp (hunk 7 above)
Comment (Mischanix) [inline lib/TerminalDisplay.cpp]: Oh, I guess the update call inside calDrawTextAdditionHeight handles this.

**Comment by agaida** [comment]
@hykwok - would you be so kind and rebase/update the PR?

**Comment by agaida** [comment]
@hykwok and @all - i played a little bit with @hykwok's patch, the behaviour would be better, but not really fixed. (I know that i be a little bit selfish, but this is happend with Noto 11pt - the font i would like to use in the terminal, 10pt and 12pt are ok :P) - shall i upload the reworked patch into a new branch?

**Comment by agaida** [comment]
applied @hoxnox  cleaned patch

**Pull request closed without merging**